### DISCUSSION TITLE
Loading JDBC driver work doesn't work from rails

### DISCUSSION THREAD

**Discussion by gaurish** [jruby/jruby#7290] [General]
I am trying load snowflake JDBC driver which works with jruby as standalone script but not within Rails. 

working as a script or jruby IRB
https://gist.github.com/gaurish/8336972c32b8645a918b4f51b4bf1c08

But the within rails, driver is not being loaded:
https://gist.github.com/gaurish/8336972c32b8645a918b4f51b4bf1c08#file-rails-console-rb

**Environment Information**
* jruby 9.3.4.0 (2.6.8) 2022-03-23 eff48c1ebf Java HotSpot(TM) 64-Bit Server VM 25.211-b12 on 1.8.0_211-b12 +jit [x86_64-darwin]
*  uname -a                                                                                                                                                         
Darwin tarantula 20.6.0 Darwin Kernel Version 20.6.0: Tue Feb 22 21:10:41 PST 2022; root:xnu-7195.141.26~1/RELEASE_X86_64 x86_64

Other relevant info you may wish to add:
* Rails 5.2
* empty rails app generated with `rails new` available at https://github.com/gaurish/debug-rails  where this issue can be reproduced.
* chat discussion with @headius https://matrix.to/#/!vyEDBdbmzqApWaugdd:matrix.org/$1649002146177698yVQJr:matrix.org?via=matrix.org&via=libera.chat&via=rrerr.net

**Expected Behavior**
* snowflake JDBC should get loaded in rails, same way it does standalone without rails

**Actual Behavior**
Errors out. Java::JavaSql::SQLException (No suitable driver found

**Comment by headius** [comment]
As noted in the chat discussion, normal IRB (non-Rails console) seems to load the driver fine, so it seems that something about the Rails console is interfering with the loading of this JDBC jar.

**Comment by headius** [comment]
I have been playing with JDBC a bit locally and I suspect this is a classloading issue.

When I use a script similar to yours to load the SQLite3 JDBC driver, it can never see the proper driver or make a connection. It only works when I specify the jar file at the command line classpath:

```
[] ~/projects/jruby $ jruby -X-C blah.rb
Unhandled Java exception: java.sql.SQLException: No suitable driver found for jdbc:sqlite::memory:
java.sql.SQLException: No suitable driver found for jdbc:sqlite::memory:
                      getConnection at java/sql/DriverManager.java:702
                      getConnection at java/sql/DriverManager.java:251
                            invoke0 at jdk/internal/reflect/NativeMethodAccessorImpl.java:-2
                             invoke at jdk/internal/reflect/NativeMethodAccessorImpl.java:62
                             invoke at jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43
                             invoke at java/lang/reflect/Method.java:566
  invokeDirectWithExceptionHandling at org/jruby/javasupport/JavaMethod.java:457
                 invokeStaticDirect at org/jruby/javasupport/JavaMethod.java:369
                             <main> at blah.rb:7
                     runInterpreter at org/jruby/Ruby.java:1291
                     runInterpreter at org/jruby/Ruby.java:1295
                        runNormally at org/jruby/Ruby.java:1180
                        runNormally at org/jruby/Ruby.java:1157
                        runNormally at org/jruby/Ruby.java:1193
                        runFromMain at org/jruby/Ruby.java:976
                      doRunFromMain at org/jruby/Main.java:396
                        internalRun at org/jruby/Main.java:280
                                run at org/jruby/Main.java:225
                               main at org/jruby/Main.java:197


[] ~/projects/jruby $ jruby -J-cp "/Users/headius/projects/jruby/lib/ruby/gems/shared/gems/jdbc-sqlite3-3.28.0/lib/sqlite-jdbc-3.28.0.jar" -X-C blah.rb
#<Java::OrgSqliteJdbc4::JDBC4Connection:0x2ba024cb>
```

I suspect that when running IRB or a script at your command line, something is causing the jar file to be loaded earlier and therefore be usable by the JDBC subsystem. When running in the Rails console, that loading does not happen.

**Comment by headius** [comment]
@gaurish A couple questions for you and things you can try:

* Do you have any environment variable that might be loading the jar file early when running the simple example?
* Please run the working script with the following JRuby flag, so we can see when the JDBC classes get loaded: `-J-verbose:class`

This has always been a frustrating process, given how JDBC loads drivers. An alternative for you would be to access the classes of the driver directly (check the docs) and request a connection directly without going through the DriverManager class.

**Comment by y-shashank** [comment]
@gaurish i was to get this working in rails by adding the driver jar file into the $CLASSPATH and using sequel gem. This is not with active-record btw

in `environment.rb` add
`$CLASSPATH << '/Users/full_path/snowflake-jdbc-3.13.16.jar'`

In order to connect 
@conn = Sequel.connect($conn_str, driver: Java::net.snowflake.client.jdbc.SnowflakeDriver)

**Comment by headius** [comment]
A second look at this makes me think that perhaps Rails is initializing the DriverManager early, so subsequent classpath updates are not seen. You could confirm this by accessing DriverManager earlier to trigger its driver discovery logic, and then run the rest of your script after that (including the `require`).

A workaround might be to have the Rails app you are running within require the jar file much earlier in the boot process, so it is available by the time it accesses DriverManager to set up the database subsystem.

**Comment by headius** [comment]
Not really a JRuby issue, I believe. DriverManager is very quirky about how and when it loads drivers, and I believe this is simply a problem with it loading before the jar file is available.

We could use better guidance here, and I'd like to get a better explanation for what is happening, but there's no bug here. I'm moving this to a discussion.

**Comment by bf4** [comment]
FWIW, I came across a similar issue and had a pretty ridiculous resolution.

The jar was in the `$CLASSPATH` and going from JRuby 9.2.21.0 to 9.4.1.0 (and Rails 6.0 without zeitwerk to 6.1 with zeitwerk) led to a new error

```
java.sql.DriverManager.getConnection(java/sql/DriverManager.java:689): No suitable driver found for jdbc:extendedsystems:advantage....(Java::JavaSql::SQLException)
        from java.sql.DriverManager.getConnection(java/sql/DriverManager.java:270)
        from java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:498)
        from org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:344)
        from org.jruby.javasupport.JavaMethod.invokeStaticDirect(org/jruby/javasupport/JavaMethod.java:256)
        from RUBY.establish_connection(/Users/benjamin/code/xbe-jdbc/lib/jdbc_adapter.rb:54)
        from RUBY.evaluate((irb):1)
        from org.jruby.RubyKernel.evalCommon(org/jruby/RubyKernel.java:1181)
        from org.jruby.RubyKernel.eval(org/jruby/RubyKernel.java:1143)
        from org.jruby.RubyKernel$INVOKER$s$0$3$eval.call(org/jruby/RubyKernel$INVOKER$s$0$3$eval.gen)
        from org.jruby.RubyKernel.loop(org/jruby/RubyKernel.java:1586)
        from org.jruby.RubyKernel$INVOKER$s$0$0$loop.call(org/jruby/RubyKernel$INVOKER$s$0$0$loop.gen)
        from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1292)
        from org.jruby.RubyKernel$INVOKER$s$rbCatch.call(org/jruby/RubyKernel$INVOKER$s$rbCatch.gen)
        from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1292)
        from org.jruby.RubyKernel$INVOKER$s$rbCatch.call(org/jruby/RubyKernel$INVOKER$s$rbCatch.gen)
```


we have a `JdbcAdapter` file which all our JDBC adapters inherit from (non active record). 

It has at the top

```ruby
java_import java.sql.Driver
java_import java.sql.DriverManager
```

and I 🆕 added at the bottom in 'main'

```ruby
# @see https://github.com/jruby/jruby/wiki/JDBC
# https://github.com/jruby/jruby/issues/7169
# https://github.com/jruby/jruby/discussions/7290
# https://github.com/jruby/jruby/issues/5742
# https://github.com/jruby/jruby/wiki/CallingJavaFromJRuby#referencing-java-classes-using-full-qualified-class-name
def reliable_java_import(fully_qualified_class_name)
  java_import fully_qualified_class_name
  instance_eval "Java::#{fully_qualified_class_name}"
end
```

then changed in a subclass


```diff
cwd = File.expand_path('..', __FILE__)
require File.join(cwd, 'jdbc_adapter')
require File.join(cwd, 'jdbc_ads', 'adsjdbc-10.10.0.28.jar')
- java_import 'com.extendedsystems.jdbc.advantage.ADSDriver'
+ reliable_java_import 'com.extendedsystems.jdbc.advantage.ADSDriver'
```

and it just works ✨ 

Weird. Thanks for the ideas.  I was working through the non-magical version (driver = Java::com.extendedsystems.jdbc.advantage.ADSDriver.new
props = java.util.Properties.new) when I discovered just invoking that made it work. 🤷
